### ISSUE TITLE
irreproducible results for resnet18 tutorial notebook

### ISSUE THREAD

**Issue by chamecall** [tianyic/only_train_once#34]
Hello, thanks for your work at first.
I wanted to run your tutorial notebook to reproduce your results (resnet18 having 0.7 sparsity and around the same accuracy) but as a result I've obtained 0.0 sparsity after 300 epochs. 
![image](https://github.com/tianyic/only_train_once/assets/51739187/ae2b0abe-1272-47be-bb21-fa07e04801b0)
I tried to change some parameters such as epsilon and start_pruning_steps but in the best case I've obtained sparsity less than 0.1 after same 300 epochs.
Сould you tell me what's the matter?
Thanks

**Comment by tianyic** [comment]
@chamecall 

Meanwhile, here is the [checkpoints](https://drive.google.com/drive/u/1/folders/1lZ7Wsehi0hr_g8nztbAFEJIhF8C4Q8Kp) by running OTOv2 onto various architectures. you could take a look or visualize via Netron. I would expect that if conducting experiments under the same dependency environment, you should be able to get similar results, some ppl told me that they may achieve better result than the tutorial shown. 

Let me know along with the training log if the issue is still there after potential dependency alignment.

**Comment by chamecall** [comment]
> @chamecall
> 
> Meanwhile, here is the [checkpoints](https://drive.google.com/drive/u/1/folders/1lZ7Wsehi0hr_g8nztbAFEJIhF8C4Q8Kp) by running OTOv2 onto various architectures. you could take a look or visualize via Netron. I would expect that if conducting experiments under the same dependency environment, you should be able to get similar results, some ppl told me that they may achieve better result than the tutorial shown.
> 
> Let me know along with the training log if the issue is still there after potential dependency alignment.

I've downgraded my onnx to 1.12.0.
Training notebook after downgrading ->
[01.quick_start_resnet18_cifar10.pdf](https://github.com/tianyic/only_train_once/files/12438043/01.quick_start_resnet18_cifar10.pdf)

My dependencies:
`addict==2.4.0
aiofiles==22.1.0
aiohttp==3.8.4
aiosignal==1.3.1
aiosqlite==0.19.0
albumentations==1.3.0
annotated-types==0.5.0
anyio==3.6.2
appdirs==1.4.4
argon2-cffi==21.3.0
argon2-cffi-bindings==21.2.0
arrow==1.2.3
asttokens==2.2.1
async-timeout==4.0.2
attrs==23.1.0
Babel==2.12.1
backcall==0.2.0
beautifulsoup4==4.12.2
better-exceptions==0.2.2
black==23.3.0
bleach==6.0.0
boto3==1.26.152
botocore==1.29.152
certifi==2022.12.7
cffi==1.15.1
charset-normalizer==3.1.0
click==8.1.3
cmake==3.25.0
coloredlogs==15.0.1
comm==0.1.3
contourpy==1.0.7
cycler==0.11.0
Cython==3.0.0
cython-bbox==0.1.3
debugpy==1.6.7
decorator==5.1.1
defusedxml==0.7.1
docker-pycreds==0.4.0
easydict==1.10
einops==0.6.1
executing==1.2.0
fastjsonschema==2.16.3
filelock==3.9.0
flatbuffers==23.5.26
fonttools==4.39.3
fqdn==1.5.1
frozenlist==1.3.3
fsspec==2023.4.0
gitdb==4.0.10
GitPython==3.1.32
graphviz==0.20.1
guppy3==3.1.3
huggingface-hub==0.14.1
humanfriendly==10.0
idna==3.4
imageio==2.28.0
importlib-metadata==6.6.0
importlib-resources==5.12.0
imutils==0.5.4
install==1.3.5
ipykernel==6.22.0
ipython==8.13.2
ipython-genutils==0.2.0
ipywidgets==8.0.6
isoduration==20.11.0
jedi==0.18.2
Jinja2==3.1.2
jmespath==1.0.1
joblib==1.2.0
json5==0.9.11
jsonpointer==2.3
jsonschema==4.17.3
jstyleson==0.0.2
jupyter==1.0.0
jupyter-console==6.6.3
jupyter-events==0.6.3
jupyter-ydoc==0.2.4
jupyter_client==8.2.0
jupyter_core==5.3.0
jupyter_server==2.5.0
jupyter_server_fileid==0.9.0
jupyter_server_terminals==0.4.4
jupyter_server_ydoc==0.8.0
jupyterlab==3.6.3
jupyterlab-pygments==0.2.2
jupyterlab-widgets==3.0.7
jupyterlab_server==2.22.1
kiwisolver==1.4.4
lap==0.4.0
lazy_loader==0.2
lightning-utilities==0.8.0
lit==15.0.7
llvmlite==0.39.1
loguru==0.7.0
markdown-it-py==2.2.0
MarkupSafe==2.1.2
matplotlib==3.7.1
matplotlib-inline==0.1.6
mdurl==0.1.2
memray==1.7.0
mistune==2.0.5
mpmath==1.2.1
multidict==6.0.4
munch==3.0.0
mypy==1.4.1
mypy-extensions==1.0.0
nbclassic==0.5.6
nbclient==0.7.4
nbconvert==7.3.1
nbformat==5.8.0
nest-asyncio==1.5.6
networkx==2.8.8
notebook==6.5.4
notebook_shim==0.2.3
numba==0.56.4
numpy==1.23.5
nvidia-cublas-cu11==11.10.3.66
nvidia-cuda-cupti-cu11==11.7.101
nvidia-cuda-nvrtc-cu11==11.7.99
nvidia-cuda-runtime-cu11==11.7.99
nvidia-cudnn-cu11==8.5.0.96
nvidia-cufft-cu11==10.9.0.58
nvidia-curand-cu11==10.2.10.91
nvidia-cusolver-cu11==11.4.0.1
nvidia-cusparse-cu11==11.7.4.91
nvidia-nccl-cu11==2.14.3
nvidia-nvtx-cu11==11.7.91
only-train-once==2.0.13
onnx==1.12.0
onnxruntime==1.15.0
opencv-python==4.7.0.72
opencv-python-headless==4.7.0.72
openvino==2023.0.0
openvino-dev==2023.0.0
openvino-telemetry==2022.3.0
packaging==23.1
pandas==2.0.1
pandocfilters==1.5.0
parso==0.8.3
pathspec==0.11.1
pathtools==0.1.2
pexpect==4.8.0
pickleshare==0.7.5
Pillow==9.3.0
platformdirs==3.4.0
plotly==5.15.0
prometheus-client==0.16.0
prompt-toolkit==3.0.38
protobuf==3.20.1
psutil==5.9.5
ptyprocess==0.7.0
pure-eval==0.2.2
pycocotools @ git+https://github.com/cocodataset/cocoapi.git@8c9bcc3cf640524c4c20a9c40e89cb6a2f2fa0e9#subdirectory=PythonAPI
pycparser==2.21
pydantic==2.0.3
pydantic_core==2.3.0
pydot==1.4.2
Pygments==2.15.1
pyparsing==3.0.9
pyrsistent==0.19.3
python-dateutil==2.8.2
python-json-logger==2.0.7
pytorch-lightning==2.0.2
pytz==2023.3
PyWavelets==1.4.1
PyYAML==6.0
pyzmq==25.0.2
qtconsole==5.4.2
QtPy==2.3.1
qudida==0.0.4
requests==2.29.0
rfc3339-validator==0.1.4
rfc3986-validator==0.1.1
rich==13.3.5
s3transfer==0.6.1
safetensors==0.3.1
scikit-image==0.20.0
scikit-learn==1.2.2
scipy==1.9.1
seaborn==0.12.2
Send2Trash==1.8.2
sentry-sdk==1.29.2
setproctitle==1.3.2
shapely==2.0.1
six==1.16.0
smmap==5.0.0
sniffio==1.3.0
soupsieve==2.4.1
stack-data==0.6.2
-e git+ssh://git@bitbucket.org/naveksoft/structured-pruning.git@8f82d994b97a75402aee39c43b84a497285ca834#egg=structured_pruning
sympy==1.11.1
tabulate==0.9.0
tenacity==8.2.2
terminado==0.17.1
texttable==1.6.7
thop==0.1.1.post2209072238
threadpoolctl==3.1.0
tifffile==2023.4.12
timm==0.9.2
tinycss2==1.2.1
tomli==2.0.1
torch==2.0.1
torch-scatter @ file:///home/algernone/Downloads/torch_scatter-2.1.1%2Bpt20cu118-cp39-cp39-linux_x86_64.whl
torchdata==0.6.1
torchmetrics==1.0.1
torchsummary==1.5.1
torchtext==0.15.2+cpu
torchvision==0.15.2
tornado==6.3.1
tqdm==4.65.0
traitlets==5.9.0
triton==2.0.0
typing_extensions==4.7.1
tzdata==2023.3
uri-template==1.2.0
urllib3==1.26.15
wandb==0.15.8
wcwidth==0.2.6
webcolors==1.13
webencodings==0.5.1
websocket-client==1.5.1
widgetsnbextension==4.0.7
y-py==0.5.9
yarl==1.9.2
-e git+https://github.com/ifzhang/ByteTrack.git@d1bf0191adff59bc8fcfeaa0b33d3d1642552a99#egg=yolox
ypy-websocket==0.8.2
zipp==3.15.0`

my device:

`+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 530.30.02              Driver Version: 530.30.02    CUDA Version: 12.1     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                  Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf            Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA GeForce RTX 3090         Off| 00000000:08:00.0 Off |                  N/A |
| 53%   38C    P8               54W / 390W|   4575MiB / 24576MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
                                                                                         
+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|    0   N/A  N/A      1439      G   /usr/lib/xorg/Xorg                          276MiB |
|    0   N/A  N/A      5709      G   ...,WinRetrieveSuggestionsOnlyOnDemand       72MiB |
|    0   N/A  N/A      6803      G   /usr/lib/firefox/firefox                    159MiB |
|    0   N/A  N/A      9491      G   ...sion,SpareRendererForSitePerProcess      107MiB |
|    0   N/A  N/A     12058      G   ...sktop/4980/usr/bin/telegram-desktop        4MiB |
|    0   N/A  N/A     98254      C   ...3.9.16/envs/python3.9/bin/python3.9     3926MiB |
|    0   N/A  N/A    116198      G   ...ures=SpareRendererForSitePerProcess       23MiB |
+---------------------------------------------------------------------------------------+
`

**Comment by tianyic** [comment]
@chamecall 

Thanks for the info. Your dependency list though is not the same as ours when we created the tutorial, looks normal for me. We are actually developing subsequent OTO-LORA upon `torch==2.0.1` though did not try RTX3090. For a trouble shooting, please try the below setup:

```python

optimizer = oto.dhspg(
    variant='sgd', 
    lr=0.1, 
    target_group_sparsity=0.7,
    weight_decay=1e-4,
    start_pruning_steps=10* len(trainloader), # start pruning after 10 epochs,
    lmbda=1e-2, # regularizer coefficient, default is 1e-3,
    epsilon=0.99)

# At the end of each epoch, log the following 4 metrics
norm_redundant, norm_important, num_groups_redundant, num_groups_important = optimizer.compute_norm_group_partitions()

```

Please let me know the training log upon the above setup.

**Comment by tianyic** [comment]
Closed as no further response. Feel free to follow up with training log upon new optimizer setup.

**Comment by chamecall** [comment]
> @chamecall
> 
> Thanks for the info. Your dependency list though is not the same as ours when we created the tutorial, looks normal for me. We are actually developing subsequent OTO-LORA upon `torch==2.0.1` though did not try RTX3090. For a trouble shooting, please try the below setup:
> 
> ```python
> optimizer = oto.dhspg(
>     variant='sgd', 
>     lr=0.1, 
>     target_group_sparsity=0.7,
>     weight_decay=1e-4,
>     start_pruning_steps=10* len(trainloader), # start pruning after 10 epochs,
>     lmbda=1e-2, # regularizer coefficient, default is 1e-3,
>     epsilon=0.99)
> 
> # At the end of each epoch, additionally log the following 4 metrics
> norm_redundant, norm_important, num_groups_redundant, num_groups_important = optimizer.compute_norm_group_partitions()
> ```
> 
> Please let me know the training log on your end upon the above setup.

@tianyic thanks lmbda set to 1e-2 has solved the problem but I'm still wondering why it takes place if there are no changes made in notebook..
